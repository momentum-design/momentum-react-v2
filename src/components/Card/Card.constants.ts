import type { Color, Height, Rounding } from './Card.types';

const CLASS_PREFIX = 'md-card';

const COLORS: Record<string, Color> = {
  INACTIVE: 'inactive',
  SUCCESS: 'success',
  TRANSPARENT: 'transparent',
};

const HEIGHTS: Record<string, Height> = {
  AUTO: 'auto',
  LARGE: 'large',
  MEDIUM: 'medium',
  SMALL: 'small',
  TINY: 'tiny',
};

const ROUNDINGS: Record<number, Rounding> = {
  0: 0,
  1: 1,
  2: 2,
};

const DEFAULTS = {
  COLOR: COLORS.INACTIVE,
  HEIGHT: HEIGHTS.SMALL,
  ROUNDING: ROUNDINGS[2],
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, COLORS, DEFAULTS, HEIGHTS, ROUNDINGS, STYLE };
