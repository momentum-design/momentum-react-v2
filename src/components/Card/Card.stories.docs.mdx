The `<Card />` component.

This component is designed to be consumed by other more complicated **Card** components.
