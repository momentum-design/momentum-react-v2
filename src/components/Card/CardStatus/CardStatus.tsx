import React, { FC } from 'react';
import classnames from 'classnames';

import { STYLE } from './CardStatus.constants';
import type { Props } from './CardStatus.types';
import './CardStatus.style.scss';

/**
 * The CardStatus component.
 */
const CardStatus: FC<Props> = (props: Props) => {
  const { className, color, id, striped, style, ...otherProps } = props;

  return (
    <div
      className={classnames(className, STYLE.wrapper)}
      data-color={color}
      data-striped={striped}
      id={id}
      style={style}
      {...otherProps}
    />
  );
};

export default CardStatus;
