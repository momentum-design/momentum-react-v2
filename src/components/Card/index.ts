import { default as Card } from './Card';
import * as CONSTANTS from './Card.constants';
import type { Props } from './Card.types';

export { CONSTANTS as CARD_CONSTANTS };

export type CardProps = Props;

export default Card;
