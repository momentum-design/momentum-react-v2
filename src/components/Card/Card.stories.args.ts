import { commonStyles, extendArgTypes } from '../../storybook/helper.stories.argtypes';

import { CARD_CONSTANTS as CONSTANTS } from './';
import { CARD_STATUS_CONSTANTS as STATUS_CONSTANTS } from './CardStatus';
import { buttonSimpleArgTypes } from '../ButtonSimple/ButtonSimple.stories.args';

const cardArgTypes = {
  children: {
    description: 'Provides the child nodes for this element.',
    control: { type: 'none' },
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  color: {
    description: 'Color of this component.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(CONSTANTS.COLORS)],
    table: {
      type: {
        summary: Object.values(CONSTANTS.COLORS)
          .map((value) => `"${value}"`)
          .join(' | '),
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.COLOR,
      },
    },
  },
  height: {
    description: 'Height of this component.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(CONSTANTS.HEIGHTS)],
    table: {
      type: {
        summary: Object.values(CONSTANTS.HEIGHTS)
          .map((value) => `"${value}"`)
          .join(' | '),
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.HEIGHT,
      },
    },
  },
  isStatic: {
    description: 'Turns off hover and active styling for the component.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  outline: {
    description: 'Whether to use an outline for this component.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  rounding: {
    description: 'Magnitude of element rounding for this component.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(CONSTANTS.ROUNDINGS)],
    table: {
      type: {
        summary: Object.values(CONSTANTS.ROUNDINGS)
          .map((value) => `${value}`)
          .join(' | '),
      },
      defaultValue: {
        summary: CONSTANTS.DEFAULTS.ROUNDING,
      },
    },
  },
  statusColor: {
    description: 'Color of the status indicator for this component.',
    control: { type: 'select' },
    options: [undefined, ...Object.values(STATUS_CONSTANTS.COLORS)],
    table: {
      type: {
        summary: Object.values(STATUS_CONSTANTS.COLORS)
          .map((value) => `${value}`)
          .join(' | '),
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
  statusStriped: {
    description: 'Whether the status indicator for this component is striped.',
    control: { type: 'boolean' },
    table: {
      type: {
        summary: 'boolean',
      },
      defaultValue: {
        summary: undefined,
      },
    },
  },
};

export { cardArgTypes };

export default {
  ...commonStyles,
  ...extendArgTypes('ButtonSimple', buttonSimpleArgTypes),
  ...cardArgTypes,
};
