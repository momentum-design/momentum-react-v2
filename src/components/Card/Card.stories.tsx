import React from 'react';

import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import Card, { CardProps, CARD_CONSTANTS as CONSTANTS } from './';
import { CARD_STATUS_CONSTANTS as STATUS_CONSTANTS } from './CardStatus';
import argTypes from './Card.stories.args';
import Documentation from './Card.stories.docs.mdx';

export default {
  title: 'Momentum UI/Card',
  component: Card,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const ChildWrapper = ({ children }: { children: string }) => (
  <div style={{ marginRight: '0.5rem', display: 'flex', alignItems: 'center' }}>{children}</div>
);

const Example = Template<CardProps>(Card).bind({});

Example.argTypes = { ...argTypes };

Example.args = {
  children: (
    <div style={{ marginRight: '0.5rem', display: 'flex', alignItems: 'center' }}>Example Text</div>
  ),
};

const Colors = MultiTemplate<CardProps>(Card).bind({});

Colors.argTypes = { ...argTypes };
delete Colors.argTypes.color;

Colors.parameters = {
  variants: Object.values(CONSTANTS.COLORS).map((color) => ({
    children: <ChildWrapper>{color}</ChildWrapper>,
    color,
  })),
};

const Heights = MultiTemplate<CardProps>(Card).bind({});

Heights.argTypes = { ...argTypes };
delete Heights.argTypes.height;

Heights.parameters = {
  variants: Object.values(CONSTANTS.HEIGHTS).map((height) => ({
    children: <ChildWrapper>{height}</ChildWrapper>,
    height,
  })),
};

const Outlines = MultiTemplate<CardProps>(Card).bind({});

Outlines.argTypes = { ...argTypes };
delete Outlines.argTypes.outline;

Outlines.parameters = {
  variants: [true, false].map((outline) => ({
    children: <ChildWrapper>{`outline = ${outline}`}</ChildWrapper>,
    outline,
  })),
};

const Roundings = MultiTemplate<CardProps>(Card).bind({});

Roundings.argTypes = { ...argTypes };
delete Roundings.argTypes.rounding;

Roundings.parameters = {
  variants: Object.values(CONSTANTS.ROUNDINGS).map((rounding) => ({
    children: <ChildWrapper>{`rounding = ${rounding}`}</ChildWrapper>,
    rounding,
  })),
};

const StatusColors = MultiTemplate<CardProps>(Card).bind({});

StatusColors.argTypes = { ...argTypes };
delete StatusColors.argTypes.statusColor;

StatusColors.parameters = {
  variants: Object.values(STATUS_CONSTANTS.COLORS).map((statusColor) => ({
    children: <ChildWrapper>{`statusColor = ${statusColor}`}</ChildWrapper>,
    statusColor,
  })),
};

const StatusStripes = MultiTemplate<CardProps>(Card).bind({});

StatusStripes.argTypes = { ...argTypes };
delete StatusStripes.argTypes.statusStriped;

StatusStripes.args = {
  statusColor: STATUS_CONSTANTS.COLORS.SUCCESS,
};

StatusStripes.parameters = {
  variants: [true, false].map((statusStriped) => ({
    children: <ChildWrapper>{`statusStriped = ${statusStriped}`}</ChildWrapper>,
    statusStriped,
  })),
};

export { Example, Colors, Heights, Outlines, Roundings, StatusColors, StatusStripes };
