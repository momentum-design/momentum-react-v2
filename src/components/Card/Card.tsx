import React, { FC } from 'react';
import classnames from 'classnames';

import ButtonSimple from '../ButtonSimple';

import CardStatus from './CardStatus';

import { DEFAULTS, STYLE } from './Card.constants';
import type { Props } from './Card.types';
import './Card.style.scss';

/**
 * The Card component.
 */
const Card: FC<Props> = (props: Props) => {
  const {
    children,
    className,
    color = DEFAULTS.COLOR,
    height = DEFAULTS.HEIGHT,
    id,
    outline,
    rounding = DEFAULTS.ROUNDING,
    statusColor,
    statusStriped,
    style,
    isStatic,
    isDisabled,
    ...otherProps
  } = props;

  return (
    <ButtonSimple
      className={classnames(className, STYLE.wrapper)}
      data-color={color}
      data-height={height}
      data-outline={outline}
      data-rounding={rounding}
      data-static={isStatic}
      data-disabled={isDisabled}
      isDisabled={isStatic || isDisabled}
      id={id}
      style={style}
      {...otherProps}
    >
      <CardStatus color={statusColor} striped={statusStriped} />
      {children}
    </ButtonSimple>
  );
};

export default Card;
