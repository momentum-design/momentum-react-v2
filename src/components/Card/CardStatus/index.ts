import { default as CardStatus } from './CardStatus';
import * as CONSTANTS from './CardStatus.constants';
import type { Props, Color } from './CardStatus.types';

export { CONSTANTS as CARD_STATUS_CONSTANTS };

export type CardStatusProps = Props;
export type CardStatusColor = Color;

export default CardStatus;
