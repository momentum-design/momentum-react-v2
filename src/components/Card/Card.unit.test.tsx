import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { BUTTON_SIMPLE_CONSTANTS } from '../ButtonSimple';

import Card, { CARD_CONSTANTS as CONSTANTS } from './';
import { CARD_STATUS_CONSTANTS as STATUS_CONSTANTS } from './CardStatus';
import { isDisabled } from '@testing-library/user-event/dist/utils';

describe('<Card />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const { container } = render(<Card />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with class name', () => {
      expect.assertions(1);

      const className = 'example-class';

      const { container } = render(<Card className={className} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const { container } = render(<Card id={id} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const { container } = render(<Card style={style} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with children', () => {
      expect.assertions(1);

      const children = <div>example child</div>;

      const { container } = render(<Card>{children}</Card>);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with color', () => {
      expect.assertions(1);

      const color = Object.values(CONSTANTS.COLORS).pop();

      const { container } = render(<Card color={color} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with height', () => {
      expect.assertions(1);

      const height = Object.values(CONSTANTS.HEIGHTS).pop();

      const { container } = render(<Card height={height} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with outline', () => {
      expect.assertions(1);

      const outline = true;

      const { container } = render(<Card outline={outline} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with rounding', () => {
      expect.assertions(1);

      const rounding = Object.values(CONSTANTS.ROUNDINGS).pop();

      const { container } = render(<Card rounding={rounding} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with statusColor', () => {
      expect.assertions(1);

      const statusColor = Object.values(STATUS_CONSTANTS.COLORS).pop();

      const { container } = render(<Card statusColor={statusColor} />);

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with statusStriped', () => {
      expect.assertions(1);

      const statusStriped = true;

      const { container } = render(<Card statusStriped={statusStriped} />);

      expect(container).toMatchSnapshot();
    });
    it('should match snapshot with isDisabled', () => {
      expect.assertions(1);

      const isDisabled = true;

      const { container } = render(<Card isDisabled={isDisabled} />);

      expect(container).toMatchSnapshot();
    });
    it('should match snapshot with isStatic', () => {
      expect.assertions(1);

      const isStatic = true;

      const { container } = render(<Card isStatic={isStatic} />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    const testid = 'test-id';

    it('should have class attribute when className prop is provided', async () => {
      expect.assertions(1);

      const className = 'example-class';

      render(<Card className={className} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.classList.contains(className)).toBe(true);
    });

    it('should have id attribute when id prop is provided', async () => {
      expect.assertions(1);

      const id = 'example-id';

      render(<Card id={id} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.id).toBe(id);
    });

    it('should have style attribute when style prop is provided', async () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      render(<Card style={style} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.style.color).toBe(style.color);
    });

    it('should have child nodes when children prop is provided', async () => {
      expect.assertions(1);

      render(
        <Card>
          <div data-testid={testid}>example-text</div>
        </Card>
      );

      const component = await screen.findByTestId(testid);

      expect(component).toBeTruthy();
    });

    it('should have data-color attribute when color prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-color';
      const color = Object.values(CONSTANTS.COLORS).pop();

      render(<Card color={color} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.getAttribute(attribute)).toBe(color);
    });

    it('should have data-height attribute when height prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-height';
      const height = Object.values(CONSTANTS.HEIGHTS).pop();

      render(<Card height={height} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.getAttribute(attribute)).toBe(height);
    });

    it('should have data-outline attribute when outline prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-outline';
      const outline = true;

      render(<Card outline={outline} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.getAttribute(attribute)).toBe(`${outline}`);
    });

    it('should have data-rounding attribute when rounding prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-rounding';
      const rounding = Object.values(CONSTANTS.ROUNDINGS).pop();

      render(<Card rounding={rounding} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.getAttribute(attribute)).toBe(`${rounding}`);
    });

    it('should have status with data-color attribute when statusColor prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-color';
      const statusColor = Object.values(STATUS_CONSTANTS.COLORS).pop();

      render(<Card statusColor={statusColor} data-testid={testid} />);

      const component = await screen.findByTestId(testid);
      const target = component.firstElementChild;

      expect(target.getAttribute(attribute)).toBe(statusColor);
    });

    it('should have status with data-striped attribute when statusStriped prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-striped';
      const statusStriped = true;

      render(<Card statusStriped={statusStriped} data-testid={testid} />);

      const component = await screen.findByTestId(testid);
      const target = component.firstElementChild;

      expect(target.getAttribute(attribute)).toBe(`${statusStriped}`);
    });

    it('should have data-static when isStatic prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-static';
      const isStatic = true;

      render(<Card isStatic={isStatic} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.getAttribute(attribute)).toBe(`${isStatic}`);
    });
    it('should have data-disabled when isDisabled prop is provided', async () => {
      expect.assertions(1);

      const attribute = 'data-disabled';
      const isDisabled = true;

      render(<Card isDisabled={isDisabled} data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.getAttribute(attribute)).toBe(`${isDisabled}`);
    });

    it('should extend ButtonSimple', async () => {
      expect.assertions(1);

      render(<Card data-testid={testid} />);

      const component = await screen.findByTestId(testid);

      expect(component.classList.contains(BUTTON_SIMPLE_CONSTANTS.STYLE.wrapper)).toBe(true);
    });
  });
});
