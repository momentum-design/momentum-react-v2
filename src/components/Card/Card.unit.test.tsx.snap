// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> snapshot should match snapshot 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with children 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
    <div>
      example child
    </div>
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with class name 1`] = `
<div>
  <button
    class="example-class md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with color 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="transparent"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with height 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="tiny"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with id 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    id="example-id"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with isDisabled 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-disabled="true"
    data-height="small"
    data-rounding="2"
    disabled=""
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with isStatic 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    data-static="true"
    disabled=""
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with outline 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-outline="true"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with rounding 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with statusColor 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
      data-color="success"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with statusStriped 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
      data-striped="true"
    />
  </button>
</div>
`;

exports[`<Card /> snapshot should match snapshot with style 1`] = `
<div>
  <button
    class="md-card-wrapper md-button-simple-wrapper"
    data-color="inactive"
    data-height="small"
    data-rounding="2"
    style="color: pink;"
    type="button"
  >
    <div
      class="md-card-status-wrapper"
    />
  </button>
</div>
`;
